### PR TITLE
feat(autodiff): add fused abc slot-recurrence scan kernel

### PR DESCRIPTION
**Pull request by ooples**
## Outcome  This PR makes ABC (Attention with Bounded-memory Control) a first-class differentiable engine operation and completes its inference path as a strict, device-resident GPU composition.  It fixes the original ABC training defect, where 5 of 9 trainable tensors received no gradient, and replaces the first GPU iteration's silent CPU fallback with observable, capability-aware execution.  ## New capabilities  - Adds AbcScanForward to the engine contract with a fused CPU forward and exact BPTT backward. - Propagates gradients through all five inputs: Q projection, K projection, V projection, forget gate, and slot keys. - Adds GPU inference without requiring a backend-specific ABC kernel. The recurrence is composed from the shared direct-backend primitive contract: device permutation/copy, scale, add/multiply, batched GEMM, softmax, slice, tile, and concatenate. - Applies the shared capability-aware precision policy introduced by #971:   - SpeedFirst keeps the caller's public Tensor<T> type while converting through the backend-selected device format.   - PreserveInputType routes to the exact CPU implementation when the backend cannot preserve T.   - Tape-active execution stays on the differentiable fused CPU node. - Supports every in-tree direct backend that implements the shared primitive and precision-capability contracts. A missing or failed primitive is reported; it is not disguised as successful GPU execution.  ## Existing behavior fixed  - ABC scalar-loop training severed autodiff upstream of the recurrence. Q/K/V projections and both forget-gate parameters never learned. - AbcScanForward caught every GPU exception and returned a CPU result, making broken device routes look successful. - The hardware test checked an inherited CpuEngine.DirectGpu property that always returned null, so it always skipped even with a working GPU. - The hardware test called the public wrapper, allowing a failed GPU composition to compare one CPU result with another and pass. - Strided compute permutations caused hidden device-to-host packing. - Recurrent gate expansion used a generic multi-axis CPU tile and caused one readback per timestep. - Intermediate recurrence tensors were left for GC instead of being disposed deterministically. - CUDA stream capture could enter an allocation-heavy recurrence and abort with CUDA 906. - numSlots == 0 failed inside softmax with a wrapped indexing exception instead of a shape error. - The global gradcheck factory and GPU dedicated-coverage registry did not recognize the new operation.  ## Type and execution semantics  | Mode | Declared type | Execution | Returned type | | --- | --- | --- | --- | | SpeedFirst | float | Backend-selected GPU format | Tensor<float> | | SpeedFirst | double | Approximate GPU path through FP32 on the tested GTX 1660 Ti | Tensor<double> | | PreserveInputType | double on a backend without FP64 ABC support | Exact CPU route | Tensor<double> | | Training / active tape | Any supported T | Fused differentiable CPU route | Tensor<T> |  The implementation is generic and delegates format selection to the backend capability planner. Physical type-policy evidence in this PR covers float and double. Reduced formats and non-CUDA devices are capability-driven but are not falsely presented here as physically tested.  ## GPU residency and performance design  - Write probabilities for the complete sequence are computed once with one batched GEMM plus one softmax. - Only the recurrent state update and slot read remain per timestep. - Compute-layout permutations are materialized device-to-device. - Gate expansion is split into supported single-axis device tiles. - The physical tests require at least one GPU launch, zero device-to-host transfers before final result materialization, and zero recorded CPU fallbacks. - Temporary tensor ownership is deterministic and bounded by the active recurrence state plus retained output steps.  This PR proves removal of host round trips; it does not claim an unmeasured wall-clock speedup for arbitrary tensor sizes.  ## Evidence  Physical device:  - NVIDIA GeForce GTX 1660 Ti, compute capability 7.5 - NVIDIA driver 610.88 - CUDA backend selected explicitly  Physical CUDA cases:  - float: [B=1, S=4, M=8, H=2, N=3] - float: [B=2, S=5, M=12, H=3, N=4] - float: [B=2, S=1, M=8, H=4, N=2], covering the no-concatenate branch - double SpeedFirst: [B=2, S=3, M=8, H=2, N=3], public double result through GPU FP32 - double PreserveInputType: [B=1, S=2, M=4, H=2, N=2], zero GPU launches and exact CPU result  Observed assertions:  - All three direct float composition cases launched device work. - SpeedFirst double launched device work and retained Tensor<double>. - Internal readbacks before result materialization: 0. - Recorded GPU-to-CPU fallbacks: 0. - SpeedFirst double versus the double CPU oracle: every element within 2e-5. - PreserveInputType double versus the CPU oracle: exact equality. - Fused CPU forward versus an independent reference: within 1e-10. - Analytic backward versus central finite differences: all five inputs covered under a non-uniform output weighting.  Post-merge test results at head a8c4975d:  - net10.0 focused ABC plus both former CI gates: 15 passed, 1 expected generated-parity skip, 0 failed. - net471 CPU/generic and gate coverage with GPU disabled: 9 passed, 5 hardware skips, 0 failed. - A forced net471 CUDA run on this CUDA-13-only host fails explicitly because .NET Framework requests cublas64_12 while only cuBLAS 13 is installed. It does not return CPU output. Therefore net471 is counted as compile/CPU evidence here, not physical GPU evidence.  ## Validation and safety  - Current with main, including #971 capability-aware autocast/type policy and #868 direct-PTX softmax work. - CUDA stream capture is rejected before input inspection or device work because this composed recurrence allocates per step. - Invalid shapes retain the canonical CPU validation surface. - numSlots must be at least one and has a dedicated regression test. - All four unresolved review threads were answered and resolved.

### CODE REVIEW

File: src/AiDotNet.Tensors/Engines/CpuEngine.AbcScan.cs [hunk 1]
```diff
@@ -0,0 +1,609 @@
...
+        if (modelDim % numHeads != 0)
+            throw new ArgumentException($"modelDim ({modelDim}) must be divisible by numHeads ({numHeads}).", nameof(numHeads));
+        int headDim = modelDim / numHeads;
+        EnsureSameShape(qProj, kProj, nameof(kProj));
+        EnsureSameShape(qProj, vProj, nameof(vProj));
+        if (forgetGate.Rank != 3 || forgetGate.Shape[0] != batch || forgetGate.Shape[1] != seqLen || forgetGate.Shape[2] != numHeads)
+            throw new ArgumentException($"forgetGate must be [batch={batch}, seqLen={seqLen}, numHeads={numHeads}].", nameof(forgetGate));
+        if (slotKeys.Rank != 3 || slotKeys.Shape[0] != numHeads || slotKeys.Shape[2] != headDim)
+            throw new ArgumentException($"slotKeys must be [numHeads={numHeads}, numSlots, headDim={headDim}].", nameof(slotKeys));
+        int numSlots = slotKeys.Shape[1];
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/CpuEngine.AbcScan.cs:69]: Fixed in 513bf79e. numSlots below one now produces an ArgumentException containing the full expected slot-key shape, and ZeroSlots_FailsWithShapeError covers the value explicitly.

File: src/AiDotNet.Tensors/Engines/DirectGpuTensorEngine.cs [hunk 2]
```diff
@@ -9951,6 +9951,124 @@ public override Tensor<T> GlaScanForward<T>(
...
+        int numSlots = slotKeys.Shape._dims[1];
+
+        try
+        {
+            return AbcScanComposed(this, qProj, kProj, vProj, forgetGate, slotKeys,
+                batch, seqLen, modelDim, numHeads, headDim, numSlots, slotInitScale);
+        }
+        catch
+        {
+            return base.AbcScanForward(qProj, kProj, vProj, forgetGate, slotKeys, numHeads, slotInitScale);
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/DirectGpuTensorEngine.cs]: Fixed in 513bf79e. CUDA stream capture is detected before shape inspection or device work. Because silent GPU-to-CPU fallback is now prohibited after backend selection, the route throws NotSupportedException instead of moving a captured inference path to the CPU.

File: src/AiDotNet.Tensors/Engines/CpuEngine.AbcScan.cs [hunk 3]
```diff
@@ -0,0 +1,609 @@
...
+    /// <param name="kProj">Key projection [batch, seqLen, modelDim].</param>
+    /// <param name="vProj">Value projection [batch, seqLen, modelDim].</param>
+    /// <param name="forgetGate">Post-sigmoid scalar-per-head forget gate [batch, seqLen, numHeads].</param>
+    /// <param name="slotKeys">Trainable slot keys [numHeads, numSlots, headDim].</param>
+    /// <param name="numHeads">Number of heads; modelDim must be divisible by it.</param>
+    /// <param name="slotInitScale">Scale applied to the slot keys to seed the initial slot state.</param>
+    /// <returns>The slot-read output [batch, seqLen, modelDim].</returns>
+    public virtual Tensor<T> AbcScanForward<T>(
+        Tensor<T> qProj, Tensor<T> kProj, Tensor<T> vProj, Tensor<T> forgetGate,
+        Tensor<T> slotKeys, int numHeads, double slotInitScale = 0.1)
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/CpuEngine.AbcScan.cs:46]: Fixed in 513bf79e (merged with current main at d5e3b6cc). Once a supported float shape selects a GPU backend, AbcScanForward enables strict primitive dispatch and backend failures now escape as a backend-specific InvalidOperationException; it no longer returns a CPU result. The composition also materializes compute permutations device-to-device and keeps tile inputs resident. On a GTX 1660 Ti, all three direct composition cases launched GPU work with zero internal readbacks and zero recorded fallbacks.

File: tests/AiDotNet.Tensors.Tests/Engines/AbcScanGpuParityTests.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+        var expected = (float[])(object)new CpuEngine()
+            .AbcScanForward(q, k, v, fg, sk, numHeads, InitScale).GetDataArray()!;
+        var got = (float[])(object)gpu
+            .AbcScanForward(q, k, v, fg, sk, numHeads, InitScale).GetDataArray()!;
+
+        Assert.Equal(expected.Length, got.Length);
+        for (int i = 0; i < expected.Length; i++)
+            Assert.True(Math.Abs(expected[i] - got[i]) < 1e-4f,
+                $"[{batch}x{seqLen}x{modelDim}, H={numHeads}, N={numSlots}] element {i}: cpu={expected[i]} gpu={got[i]}");
```
[the remark this answers is not in the document]
Comment (ooples) [reply tests/AiDotNet.Tensors.Tests/Engines/AbcScanGpuParityTests.cs]: Fixed in 513bf79e. The hardware test now calls AbcScanComposed directly, uses the actual IsGpuAvailable property (the former inherited DirectGpu check always returned null and always skipped), and asserts launches greater than zero, readbacks equal zero, and no fallbacks before output materialization. All three cases passed on a GTX 1660 Ti; the global dedicated-coverage gate also passes.

File: src/AiDotNet.Tensors/Engines/DirectGpuTensorEngine.cs [hunk 5]
```diff
@@ -10022,6 +10022,220 @@ public override Tensor<T> GlaScanForward<T>(
...
+        catch (Exception ex)
+        {
+            throw new InvalidOperationException(
+                $"AbcScanForward failed on the selected {abcBackend.DeviceType} backend; CPU fallback is disabled for this GPU route.",
+                ex);
+        }
+        finally
+        {
+            ThrowOnGpuKernelFallback = previousThrowOnFallback;
+        }
```
[the remark this answers is not in the document]
Comment (ooples) [reply src/AiDotNet.Tensors/Engines/DirectGpuTensorEngine.cs:10101]: Fixed in a8c4975d. AbcScanForward now lets OutOfMemoryException and OperationCanceledException propagate unchanged. Strict mode is also enforced at the Softmax, TensorAdd, TensorMultiply, TensorMultiplyScalar, TensorConcatenate, and TensorSliceAxis CPU exits. The ABC GPU composition itself now uses direct resident helpers for add, multiply, scale, slice, tile, permute, and concatenate, so those operations do not traverse public fallback branches. Post-fix validation: net10 focused ABC 13 passed / 1 expected skip; both global gates passed; net471 library build succeeded with zero warnings and errors.
